apps/aechannel/src/aesc_fsm.erl

```diff
@@ -1376,23 +1374,113 @@ new_onchain_tx(channel_withdraw_tx, #{acct := ToId,
+                                responder := R}
+                    , state = State } = D, CurrHeight) ->
+    Account = my_account(D),
+    Def = tx_defaults(channel_settle_tx, #{acct => Account}, D),
+    Opts1 = maps:merge(Def, Opts),
+    ChanId = maps:get(channel_id, Opts1),
+    {ok, IAmt} = aesc_offchain_state:balance(I, State),
+    {ok, RAmt} = aesc_offchain_state:balance(R, State),
+    TTL = adjust_ttl(maps:get(ttl, Opts1, 0)),
+    SlashOpts = Opts1#{ channel_id => aeser_id:create(channel, ChanId)
+                      , from_id    => aeser_id:create(account, Account)
+                      , initiator_amount_final => IAmt
+                      , responder_amount_final => RAmt
+                      , ttl        => TTL},
+    {ok, Tx} = new_onchain_tx_(aesc_settle_tx, SlashOpts, CurrHeight),
+    {ok, Tx, []};
+new_onchain_tx(channel_close_solo_tx, Opts,
+               #data{ on_chain_id = ChanId
+                    , opts = #{initiator := Initiator,
+                               responder := Responder}
+                    , state       = State } = D, CurrHeight) ->
+    Account = my_account(D),
+    TTL = adjust_ttl(maps:get(ttl, Opts, 0)),
+    {_Round, SignedTx} = aesc_offchain_state:get_latest_signed_tx(State),
+    Payload = case aesc_utils:is_offchain_tx(SignedTx) of
+                  true ->
+                      aetx_sign:serialize_to_binary(SignedTx);
+                  false ->
+                      %% on-chain tx, we assume included in the chain
+                      <<>>
+              end,
+    {ok, Poi} = aesc_offchain_state:poi([{account, Initiator},
+                                         {account, Responder}], State),
+    Opts1 =
+          Opts#{ channel_id => aeser_id:create(channel, ChanId)
+               , from_id    => aeser_id:create(account, Account)
+               , payload    => Payload
+               , poi        => Poi
+               , ttl        => TTL},
+    {ok, Tx} = new_onchain_tx_(aesc_close_solo_tx, Opts1, CurrHeight),
+    {ok, Tx, []};
+new_onchain_tx(channel_slash_tx, Opts,
+               #data{ on_chain_id = ChanId
+                    , opts = #{initiator := Initiator,
+                               responder := Responder}
+                    , state       = State} = D, CurrHeight) ->
+    Account = my_account(D),
+    Def = tx_defaults(channel_slash_tx, #{acct => Account}, D),
+    Opts1 = maps:merge(Def, Opts),
+    TTL = adjust_ttl(maps:get(ttl, Opts1, 0)),
+    {ok, Poi} = aesc_offchain_state:poi([{account, Initiator},
+                                         {account, Responder}], State),
+    Opts2 =
+        Opts1#{ channel_id => aeser_id:create(channel, ChanId)
+              , from_id    => aeser_id:create(account, Account)
+              , poi        => Poi
+              , ttl        => TTL},
+    {ok, Tx} = new_onchain_tx_(aesc_slash_tx, Opts2, CurrHeight),
+    {ok, Tx, []}.
+
+
+
+
+new_onchain_tx_(Mod, Opts, CurrHeight) ->
```

Comment:
This should have a guard making sure that `Mod` is a supported value. That way we will find out at compile-time and test-time if we use it wrong.